github/githubapp.go

```diff
@@ -0,0 +1,109 @@
+package github
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"path"
+	"strconv"
+	"time"
+
+	"github.com/golang-jwt/jwt/v4"
+)
+
+type GitHubApp struct {
+	ClientId       string `json:"client_id"`
+	InstallationId int64  `json:"installation_id"`
+	PrivateKey     string `json:"private_key"` // SENSITIVE
+}
+
+func (app *GitHubApp) IsZero() bool {
+	if app.ClientId != "" {
+		return false
+	}
+	if app.InstallationId != 0 {
+		return false
+	}
+	if app.PrivateKey != "" {
+		return false
+	}
+	return true
+}
+
+// generateJWTtoken returns a signed JWT token used to authenticate as GitHub App
+func generateJWTtoken(clientId, privateKey string) (string, error) {
+	key, err := jwt.ParseRSAPrivateKeyFromPEM([]byte(privateKey))
+	if err != nil {
+		return "", fmt.Errorf("could not parse private key: %w", err)
+	}
+	// GitHub rejects expiry and issue timestamps that are not an integer,
+	// while the jwt-go library serializes to fractional timestamps.
+	// Truncate them before passing to jwt-go.
+	// Additionally, GitHub recommends setting this value 60 seconds in the past.
+	iat := time.Now().Add(-60 * time.Second).Truncate(time.Second)
+	// maximum validity 10 minutes. Here, we reduce it to 2 minutes.
+	exp := iat.Add(2 * time.Minute)
+	// Docs: https://docs.github.com/en/apps/creating-github-apps/authenticating-with-a-github-app/generating-a-json-web-token-jwt-for-a-github-app#about-json-web-tokens-jwts
+	claims := &jwt.RegisteredClaims{
+		IssuedAt:  jwt.NewNumericDate(iat),
+		ExpiresAt: jwt.NewNumericDate(exp),
+		// The client ID or application ID of your GitHub App.
+		// Use of the client ID is recommended.
+		Issuer: clientId,
+	}
+
+	// GitHub JWT must be signed using the RS256 algorithm.
+	token, err := jwt.NewWithClaims(jwt.SigningMethodRS256, claims).SignedString(key)
+	if err != nil {
+		return "", fmt.Errorf("could not sign the JWT token: %w", err)
+	}
+	return token, nil
+}
+
+// GenerateInstallationToken returns an installation token used to authenticate as GitHub App installation
+func GenerateInstallationToken(server string, app GitHubApp) (string, error) {
+	// API: POST /app/installations/{installationId}/access_tokens
+	installationId := strconv.FormatInt(app.InstallationId, 10)
+	url := server + path.Join("/app/installations", installationId, "/access_tokens")
+
+	req, err := http.NewRequest(http.MethodPost, url, nil)
+	if err != nil {
+		return "", fmt.Errorf("github post: new request: %s", err)
+	}
+	req.Header.Add("Accept", "application/vnd.github.v3+json")
+
+	jtwToken, err := generateJWTtoken(app.ClientId, app.PrivateKey)
+	if err != nil {
+		return "", err
+	}
+	req.Header.Set("Authorization", "Bearer "+jtwToken)
+
+	client := &http.Client{Timeout: time.Second * 5}
```

Comment:
Here we care about the timeout, but we have line 70 before.
I think that we can share the client, it might even keep the TCP connection open.

Reply:
Yes.. This is why I wrote the following comment in the original PR. See: https://github.com/Pix4D/cogito/pull/161/files#r2039756094

But, I think this would be a better fit to a separate PR... If you wish, I can first to a refactoring PR and sort this properly...

Reply:
Note also that we have the same issues in github commit status 
https://github.com/Pix4D/cogito/blob/67e7fc9da35caef9c6c461eb265968734e23d880/github/commitstatus.go#L137

while in gchat sink we have request that can be terminated by context, but client with no timeout...
https://github.com/Pix4D/cogito/blob/67e7fc9da35caef9c6c461eb265968734e23d880/googlechat/googlechat.go#L99

Reply:
> while in gchat sink we have request that can be terminated by context, but client with no timeout...

the timeout in the http client struct has been added before "context", so if one is using "context", no need to set the timeout, and, only partially, vice-versa (only partially because "context" has effect on more that the timeout of the http client struct).

Reply:
See: https://github.com/Pix4D/cogito/pull/165